### PR TITLE
Split XDebug into it's own container so it doesn't slow normal use

### PR DESCRIPTION
**Pull request by charrondev**
Just **_having_** the xdebug extension installed causes a big slowdown to PHP, so I've done some optimization. We now have 2 php containers that are dynamically routed to with nginx.

## Base PHP container

This is just like our previous php container, but crucially **_doesn't have XDebug_**. As a result, it's much, much faster.

Look at the difference of running the container with and without XDebug

image

I've also added OPCache with a development configuration to speed things up a bit in this container.

## XDebug container

This one takes the new base container and adds xdebug. It mounts to a different socket than the normal PHP so nginx can distinguish them.

## How it knows?

Any of the following will cause nginx to send you to the XDebug extension instead of the normal one.

- URL has a query param `XDEBUG_PROFILE=1` (for profiling).
- URL has a query param `XDEBUG_SESSION_START=1` or `XDEBUG_SESSION_START=phpstorm`.
- Request has an XDEBUG_SESSION cookie.

I've also reduced the cookie time to 1 second so you can get back to normal quickly after a debugging session. I'd recommend using a browser plugin to keep setting the cookie or to apply the URL when you want debugging.

### CODE REVIEW

**Review by initvector — CHANGES_REQUESTED** [review]
I haven't tested the functionality, yet, but it sounds very promising! I only have a few minor questions and suggestions on the code, itself.

File: docker-compose.unit-tests.yml [hunk 1]
```diff
@@ -5,23 +5,23 @@ services:
     unit-test-php-fpm:
         build:
             context: "./images/php-fpm"
+            dockerfile: DockerfileXDebug
         container_name: "unit-test-php-fpm"
         environment:
-            - "PHP_INI_SCAN_DIR=:/usr/local/etc/php/custom.conf.d" # Allow users to override configs
```
[the remark this answers is not in the document]
Comment (charrondev) [reply docker-compose.unit-tests.yml:10]: People who want that can add them to the main ini files. It's confusing having stuff in multiple places, and we don't want to encourage long term usage of custom files.

If you have a custom config, you should make a PR and get it upstreamed. The whole purpose of the docker setup here is keep people running the same setup.

[the remark this answers is not in the document]
Comment (initvector) [reply docker-compose.unit-tests.yml:10]: I'm onboard with removing this functionality (don't think I've ever actually used it), but I didn't see it mentioned in the PR description, so I wasn't sure if its addition was intentional. 👍

File: images/php-fpm/Dockerfile [hunk 2]
```diff
@@ -3,46 +3,44 @@ FROM php:7.4-fpm-buster
...
+    libjpeg62-turbo-dev \
+    libpng-dev \
+    # for mbstring
+    libonig-dev \
+    # memchached extension
+    libmemcached-dev \
+    # allow mailing to work
+    sendmail \
+    # allow reading of image exif
+    exiftool \
```
[the remark this answers is not in the document]
Comment (charrondev) [reply images/php-fpm/Dockerfile:19]: It was my autoformatter. Fixing it.

File: images/php-fpm/docker-entrypoint.sh [hunk 3]
```diff
@@ -1,5 +1,7 @@
 #!/bin/bash
 
+# Symlink extra xdebug stuff in required, otherwise deleted
```
[the remark this answers is not in the document]
Comment (charrondev) [reply images/php-fpm/docker-entrypoint.sh:3]: Unneeded. Left over from something previous.

File: docker-compose.yml [hunk 4]
```diff
@@ -63,24 +63,30 @@ services:
     php-fpm:
         build:
             context: "./images/php-fpm"
+            dockerfile: Dockerfile
```
[the remark this answers is not in the document]
Comment (charrondev) [reply docker-compose.yml:66]: Done.

File: docker-compose.unit-tests.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (charrondev) [reply docker-compose.unit-tests.yml:10]: :y: I don't know of anyone that's used it yet.

**Review by initvector — APPROVED** [review]
🎸

**Pull request merged**